packages/editor/src/browser/navigation/navigation-location-service.ts

```diff
@@ -233,4 +235,45 @@ ${this.stack.map((location, i) => `${i}: ${JSON.stringify(NavigationLocation.toO
 ----- o -----`;
     }
 
+    // A history of recently closed editors.
+    private recentlyClosedEditors: EditorWidget[] = [];
+    private static readonly MAX_RECENTLY_CLOSED_EDITORS = 30;
+
+    /**
+     * Appends new closed editor to the history.
+     * Removes old editor that is duplicated of the given editor if exists.
+     * Removes old editors if the number of editors in the history exceeds MAX_RECENTLY_CLOSED_EDITORS.
+     *
+     * @param editor The newly closed editor.
+     */
+    pushToRecentlyClosedEditors(editor: EditorWidget): void {
+        this.removeFromRecentlyClosedEditors(editor);
+        this.recentlyClosedEditors.push(editor);
+        if (this.recentlyClosedEditors.length > NavigationLocationService.MAX_RECENTLY_CLOSED_EDITORS) {
+            this.recentlyClosedEditors.shift();
+        }
+    }
+
+    /**
+     * Removes the given editor if it exists in closed editors history.
+     *
+     * @param target The editor that should be removed from history.
+     */
+    private removeFromRecentlyClosedEditors(target: EditorWidget): void {
+        this.recentlyClosedEditors = this.recentlyClosedEditors.filter(e => {
+            if (target.editor.uri.isEqual(e.editor.uri)) {
+                return false;
+            }
+            return true;
+        });
+    }
+
+    /**
+     * Removes the last closed editor out of the history.
+     *
+     * @returns The last closed editor.
+     */
+    popFromRecentlyClosedEditors(): EditorWidget | undefined {
+        return this.recentlyClosedEditors.pop();
+    }
```

Comment:
A better name should be provided for clients, `pop` is an implementation detail and not necessarily meaningful for clients. If you want to keep `pop` (for documentation purposes that we remove the editor from the history), create two methods, the private `pop`, and a public `getter` for clients.

```suggestion
    /**
     * Get the last recently closed editor.
     *
     * @returns the last recently closed editor if it exists.
     */
    getLastRecentlyClosedEditor(): EditorWidget | undefined {
        return this.popFromRecentlyClosedEditors(); // or return this.recentlyClosedEditors.pop();
    }
```